perllib/FixMyStreet/Cobrand/Bexley/Waste.pm

```diff
@@ -1404,44 +1515,88 @@ sub waste_munge_request_data {
+
+    if ( $type eq 'delivery' ) {
+        my $delivery_containers
+            = $c->stash->{property}{containers_for_delivery};
+
+        for my $dc (@$delivery_containers) {
+            $service = $dc if $id eq ( $dc->{service_item_name} // '' );
+
+            last if $service;
+
+            if ( @{ $dc->{subtypes} // [] } ) {
+                # The service we are looking for may be the subtype of
+                # a parent
+                # (e.g. 'RES-140' under 'Green Wheelie Bin')
+                for my $subtype ( @{ $dc->{subtypes} } ) {
+                    if ( $id eq $subtype->{service_item_name} ) {
+                        $service = $subtype;
+                        # Use parent name
+                        $service->{name} = $dc->{name};
+                        last;
+                    }
                 }
             }
+
+            last if $service;
+
+            # Some containers have an original service_item_name (ID) with spaces.
+            # E.g. 'Deliver Box lids 55L'.
+            # We need to unhyphen string that was hyphenated in
+            # construct_bin_request_form().
+            my $id_spaced = $id =~ s/-/ /gr;
+
+            if ( $id_spaced eq ( $dc->{service_item_name} // '' ) ) {
+                $service = $dc;
+            }
+
+            last if $service;
         }
 
-        last if $service;
+        $data->{title}    = "Request new $service->{name}";
+        $data->{category} = $data->{category_delivery};
 
-        # Some containers have an original service_item_name (ID) with spaces.
-        # E.g. 'Deliver Box lids 55L'.
-        # We need to unhyphen string that was hyphenated in
-        # construct_bin_request_form().
-        my $id_spaced = $id =~ s/-/ /gr;
+    } else {
+        my $removal_containers
+            = $c->stash->{property}{containers_for_removal};
+
+        my $original_id = $id =~ s/-removal$//r;
+
+        for my $rc (@$removal_containers) {
+            $service = $rc
+                if $original_id eq ( $rc->{service_item_name} // '' );
+
+            last if $service;
+
+            if ( @{ $rc->{subtypes} // [] } ) {
+                # The service we are looking for may be the subtype of
+                # a parent
+                # (e.g. 'RES-140' under 'Green Wheelie Bin')
+                for my $subtype ( @{ $rc->{subtypes} } ) {
+                    if ( $original_id eq $subtype->{service_item_name} ) {
+                        $service = $subtype;
+                        # Use parent name
+                        $service->{name} = $rc->{name};
+                        last;
+                    }
+                }
+            }
 
-        if ( $id_spaced eq ( $dc->{service_item_name} // '' ) ) {
-            $service = $dc;
+            last if $service;
         }
 
-        last if $service;
-    }
+        $data->{title}    = "Request removal of $service->{name}";
+        $data->{category} = $data->{category_removal};
 
```

Comment:
This block is almost identical to the request block just above if we want to factor those together